article.md

```diff
@@ -0,0 +1,265 @@
+```cairo
+@external
+func add_connection{
+    syscall_ptr : felt*,
+    pedersen_ptr : HashBuiltin*,
+    range_check_ptr,
+    bitwise_ptr : BitwiseBuiltin*
+}(eth_address : felt, starknet_address : felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    assert_valid_eth_signature(eth_address, starknet_address, domain_hash, signature_len, signature);
+    save_connected_addresses(eth_address, starknet_address);
+    return ();
+}
+```
+
+The hardest part for us is getting the hash of our message. Once we have that, we can just use a built-in function `verify_eth_signature_uint256` which will do all the heavy lifting, dealing with complicated math, for us. 
+So we will have to write the `get_hash` function for ourselves.
+```cairo
+func assert_valid_eth_signature{
+        syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr,
+        bitwise_ptr : BitwiseBuiltin*
+}(eth_address: felt, starknet_address: felt, domain_hash: Uint256, signature_len: felt, signature: felt*){
+    alloc_locals;
+    with_attr error_message(
+        "Invalid signature length. Signature should have exactly 5 elements."
+    ){
+        assert signature_len = 5;
+    }
+
+    let v = signature[0];
+    let r = Uint256(signature[2], signature[1]);
+    let s = Uint256(signature[4], signature[3]);
+    
+    let (local keccak_ptr_start) = alloc();
+    let keccak_ptr = keccak_ptr_start;
+
+    let (hash_uint) = get_hash{keccak_ptr=keccak_ptr}(starknet_address, domain_hash);
+   
+    verify_eth_signature_uint256{keccak_ptr=keccak_ptr}(hash_uint, r, s, v, eth_address);
+
+    finalize_keccak(keccak_ptr_start=keccak_ptr_start, keccak_ptr_end=keccak_ptr);
+
+    return ();
+}
+```
+## Recreating and hashing the message
+### Constants
+Instead of calculating the structure hash every time, we can precalculate it and store it as a constant to save us some difficult operations. Additionally, Ethereum uses a pre-chosen prefix, which is explained in the [documentation](https://eips.ethereum.org/EIPS/eip-712)
+
+```cairo
+const PREFIX = 0x1901;
+
+// A Hash of a structure that we're signing (Payload) which is shown in struct.json file
+const TYPE_HASH_HIGH = 0xd3edf21d0254954db14d94abab56644c;
+const TYPE_HASH_LOW = 0x1100d60cff7b050ffcb29574618d516e;
+```
+Because we are using keccak256 for hashing, we need to store the result in two pieces (lower and higher 128 bits) as felts store only up to 251 bits. <br/><br/>
+We can do the same with the domain separator, but there is one small obstacle - we'll be using this contact's address as a variable in the domain, and we won't know this address until it's deployed, so how can we deal with it? We certainly don't need to calculate it every time the function is called; we can just create a storage variable that can only be set once and put our precalculated domain separator there after contract deployment.
```

Comment:
We **could** do...

", and we won't know this address until it's deployed" -> and hardcoding this value would affect resulting hash, changing the value itself.